replacy/__init__.py

```diff
@@ -136,28 +143,48 @@ def get_predicates(self, match_hooks) -> List[Callable]:
             predicates.append(pred)
         return predicates
 
-    def inflect_suggestion(self, pre_suggestion, doc, start, end, match_name):
+    def process_suggestions(self, pre_suggestion, doc, start, end, match_name):
         """
+        Perform inflection and replace references to the matched token
         example of pattern and pre_suggestion
         pattern: "LEMMA": "chock", "TEMPLATE_ID": 1
         pre_suggestion: "TEXT": "chalk", "FROM_TEMPLATE_ID": 1
         inflect suggestion "chalk" according to form of "cholk" from patterns
         """
         text_list = []
         for item in pre_suggestion:
-            text = item["TEXT"]
+            try:
+                text = item["TEXT"]
+            except KeyError:
+                ref = int(item["PATTERN_REF"])
+                refd_token = doc[start + ref]
             changed_text = None
 
             # check if inflect
             if "FROM_TEMPLATE_ID" in item:
                 template_id = item["FROM_TEMPLATE_ID"]
                 index = None
-                for i, token in enumerate(self.match_dict[match_name]["patterns"]):
-                    if "TEMPLATE_ID" in token and token["TEMPLATE_ID"] == template_id:
+                for i, token in enumerate(
+                    self.match_dict[match_name]["patterns"]
+                ):
+                    if (
+                        "TEMPLATE_ID" in token
+                        and token["TEMPLATE_ID"] == template_id
+                    ):
                         index = i
                         break
                 if index is not None:
-                    changed_text = self.inflector.inflect(doc, text, start + index)
+                    changed_text = self.inflector.auto_inflect(
+                        doc, text, start + index
+                    )
+            elif "PATTERN_REF" in item:
+                if "INFLECTION" in item:
+                    form = item["INFLECTION"]
+                    changed_text = self.inflector.inflect_or_lookup(
+                        refd_token, form
+                    )
+                else:
+                    changed_text = refd_token.text
```

Comment:
Wondering whether we should use `text_with_ws` instead. Token text can drop ws, unlike `TEXT`.
Fix in `inflect_or_lookup` would be needed too.